innovation_sweet_spots/pipeline/pilot/investment_predictions/create_dataset/utils.py

```diff
@@ -721,3 +684,246 @@ def add_founders(cb_data: pd.DataFrame, org_id_founders: pd.DataFrame) -> pd.Dat
+            }
+        )
+        .assign(ukri_grant=0)
+        .drop(columns="org_name")
+    )
+
+
+def explode_cb_gtr_lookup(cb_gtr_lookup: pd.DataFrame) -> pd.DataFrame:
+    """Explode crunchbase to gtr organisation lookup"""
+    # Convert gtr_org_ids type to string
+    cb_gtr_lookup["gtr_org_ids"] = cb_gtr_lookup["gtr_org_ids"].apply(literal_eval)
+    # Explode cb to gtr lookup
+    return cb_gtr_lookup.explode("gtr_org_ids").reset_index(drop=True)
+
+
+def add_gtr_project_to_cb_gtr_lookup(
+    cb_gtr_org_lookup: pd.DataFrame, gtr_grants: pd.DataFrame
+) -> pd.DataFrame:
+    """Add gtr project information to cb_gtr_org_lookup"""
+    return (
+        cb_gtr_org_lookup.merge(
+            right=gtr_grants, left_on="gtr_org_ids", right_on="gtr_org_id", how="left"
+        )
+        .dropna(subset=["gtr_org_id"])
+        .drop(columns="gtr_org_ids")
+        .drop_duplicates(subset=["project_id", "cb_org_id"])
+        .reset_index(drop=True)
+    )
+
+
+def standardise_gtr_grants(cb_gtr_lookup_projects):
+    """Standardise gtr grants data so that it can be
+    combined with crunchbase grants data"""
+    return (
+        cb_gtr_lookup_projects.rename(
+            columns={
+                "project_id": "grant_id",
+                "fund_start": "grant_date",
+                "amount": "grant_amount_gbp",
+            }
+        )
+        .assign(ukri_grant=1)
+        .drop(columns="gtr_org_id")
+    )
+
+
+def join_cb_gtr_grants(
+    cb_grants_standardised: pd.DataFrame, gtr_grants_standardised: pd.DataFrame
+) -> pd.DataFrame:
+    """Join crunchbase and gtr grants together"""
+    return pd.concat([cb_grants_standardised, gtr_grants_standardised]).reset_index(
+        drop=True
+    )
+
+
+def grants_features(combined_cb_gtr_grants: pd.DataFrame) -> pd.DataFrame:
+    """Create groupby grants features for each crunchbase organisation id"""
+    return (
+        combined_cb_gtr_grants.fillna({"grant_amount_gbp": 0})
+        .groupby("cb_org_id")
+        .agg(
+            total_grant_amount_gbp=("grant_amount_gbp", "sum"),
+            n_grants=("grant_id", "count"),
+            first_grant_date=("grant_date", "min"),
+            last_grant_date=("grant_date", "max"),
+            has_received_ukri_grant=("ukri_grant", "max"),
+        )
+        .assign(has_received_grant=1)
+        .reset_index()
+    )
+
+
+def last_grant_amount(combined_cb_gtr_grants: pd.DataFrame) -> pd.DataFrame:
+    """Calculate last grant amount received for each crunchbase organisation id"""
+    return (
+        combined_cb_gtr_grants.fillna({"grant_amount_gbp": 0})
+        .sort_values("grant_date", ascending=False)
+        .groupby("cb_org_id")
+        .head(1)
+        .rename(
+            columns={
+                "grant_date": "last_grant_date",
+                "grant_amount_gbp": "last_grant_amount_gbp",
+            }
+        )[["cb_org_id", "last_grant_amount_gbp"]]
+    )
+
+
+def gtr_projects_with_lead_orgs(
```

Comment:
This is nice - so we could in principle (much later down the line) have a feature reflecting what kind of collaborations are happening in the project (eg University + Company vs just Company)?

Reply:
Yes, that's a nice idea. I think we could have a a feature looking at the types of collaborations.